File: day-06-ruby/part_two.rb
# http://adventofcode.com/2017/day/6

require "./part_one"

def solve(blocks)
  seen = {}
  cycles = 0

  while seen[blocks] == nil do
    seen[blocks] = cycles
    blocks = reallocate blocks
    cycles += 1
  end

  cycles - seen[blocks]
end

File: 2017/day-15-javascript/PartTwo.js
// http://adventofcode.com/2017/day/15

let { areGeneratorsMatching } = require("./PartOne");

function* createGenerator(seed, multipleOf, factor) {
  let value = seed;

  while (true) {
    value = (value * factor) % 2147483647;
    if (value % multipleOf === 0) {
      yield value;
    }
  }
}

function solve(a, b) {
  let generatorA = createGenerator(a, 4, 16807);
  let generatorB = createGenerator(b, 8, 48271);
  let matchedPairCount = 0;

  for (let i = 0; i < 5000000; i++) {
    if (areGeneratorsMatching(generatorA, generatorB)) {
      matchedPairCount += 1;
    }
  }

  return matchedPairCount;
}

